augur/filter/subsample.py

```diff
@@ -297,6 +299,69 @@ def get_probabilistic_group_sizes(groups, target_group_size, random_seed=None):
     return max_sizes_per_group
 
 
+def get_weighted_group_sizes(groups, group_by, weights_file, target_total_size):
+    """Return group sizes based on weights defined in ``weights_file``.
+
+    Returns
+    -------
+    dict :
+        Mapping between groups (combinations of grouping column values in tuple
+        form) to group sizes
+    """
+    weights = pd.read_csv(weights_file, delimiter='\t')
+
+    weighted_columns = get_weighted_columns(weights_file)
+
+    # Allow other columns in group_by to be equally weighted (uniform sampling)
+    unweighted_columns = list(set(group_by) - set(weighted_columns))
+
+    if unweighted_columns:
```

Comment:
Great example! I've pushed it as a test in 740e74883e0d361d08cec97eaffd1560d365106d with an edit to create a metadata file with 250 sequences so `--subsample-max-sequences 100` makes sense.

I think the behavior is expected, and related to what the observation in your [other thread](https://github.com/nextstrain/augur/pull/1454#discussion_r1596227089). I tried to explain it here, but maybe we can walk through some more scenarios on a call.

https://github.com/nextstrain/augur/blob/740e74883e0d361d08cec97eaffd1560d365106d/tests/functional/filter/cram/subsample-weighted-and-uniform-mix.t#L60-L62

Reply:
> I think the behavior is expected

I don't agree. Let me use a more extreme (but not unheard of) example. The temporal distribution of samples across geographies can vary greatly - sometimes they're only available for one single year, sometimes every year for decades. Using the same command and weights as before with the following metadata (but imagining for each row there's many entries with the same `{date,location}` metadata so we have plenty of samples to choose from):

```console
strain	date	location
SEQ_A.1	2000	A
SEQ_A.2	2000	A
SEQ_A.3	2000	A
SEQ_B.1	2000	B
SEQ_B.2	2001	B
SEQ_B.3	2002	B
SEQ_B.4	2003	B
SEQ_B.5	2004	B
SEQ_B.6	2005	B
SEQ_B.7	2006	B
```

If I request `--group-by location year` I want around 67 x location A and 33 x location B because of the 2:1 weighting I specified. What I'll actually get is ~22 from location A and 77 from location B. The more unequal the temporal distribution of data is across locations the more the output will deviate from my specified weightings.
 
Put another way, I think extra group-bys should be used to partition the data _within_ the existing weighted groups, and thus the temporal distribution of data from a location will only affect the sampling of data from that location not other locations.

Reply:
> Put another way, I think extra group-bys should be used to partition the data _within_ the existing weighted groups, and thus the temporal distribution of data from a location will only affect the sampling of data from that location not other locations.

This makes sense. I'll look into updating the logic to reflect this.

Reply:
Done in 1a9dd7cadbbe899464e94b1aee2cfe7e0bd6d66f. Gonna test on https://github.com/nextstrain/ncov/pull/1106 tomorrow.